component/discovery/file/file.go

```diff
@@ -0,0 +1,346 @@
+			c.mut.Lock()
+			err := c.watcher.Close()
+			if err != nil {
+				level.Error(c.opts.Logger).Log("msg", "error closing watcher", "err", err)
+			}
+			c.mut.Unlock()
+			return nil
+		}
+	}
+}
+
+// reconcileWatchesWithWatcher checks for any new directories that have been added along with verifying
+// that the args and watchers are in sync.
+func (c *Component) reconcileWatchesWithWatcher() {
+	includedPaths := c.args.getPaths()
+	excludedPaths := c.args.getExcluded()
+	expandedPaths, err := getPaths(includedPaths)
+	if err != nil {
+		level.Error(c.opts.Logger).Log("msg", "error expanding paths", "err", err)
+		return
+	}
+	alreadyWatching := c.watcher.WatchList()
+	alreadyWatchingDir := make(map[string]struct{})
+	for _, p := range alreadyWatching {
+		alreadyWatchingDir[p] = struct{}{}
+	}
+	// Ensure all the paths are added.
+	for _, n := range expandedPaths {
+		fi, fileErr := os.Stat(n)
+		if fileErr != nil {
+			level.Error(c.opts.Logger).Log("msg", "error getting os stats", "err", err)
+			continue
+		}
+		if fi.IsDir() {
+			// Check to see if we are already watching.
+			if _, found := alreadyWatchingDir[n]; found {
+				continue
+			}
+			err = c.watcher.Add(n)
+			if err != nil {
+				level.Error(c.opts.Logger).Log("msg", "error adding path to watcher", "err", err)
+			}
+		} else {
+			exclude := false
+			for _, excluded := range excludedPaths {
+				if match, _ := doublestar.Match(excluded, n); match {
+					exclude = true
+					break
+				}
+			}
+			if exclude {
+				continue
+			}
+			c.addToWatchedFiles(n)
+			dir := filepath.Dir(n)
+			err = c.watcher.Add(dir)
+			if err != nil {
+				level.Error(c.opts.Logger).Log("msg", "error adding path to watcher", "err", err)
+			}
+		}
+	}
+	// Find all the removed paths.
+	filesToRemove := make([]string, 0)
+	for p := range c.watchedFiles {
+		found := false
+		for _, np := range expandedPaths {
+			if p == np {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			filesToRemove = append(filesToRemove, p)
+		}
+		// Scan to see if we need to exclude any new files.
+		for _, exclude := range excludedPaths {
+			matched, _ := doublestar.PathMatch(exclude, p)
+			if matched {
+				filesToRemove = append(filesToRemove, p)
+			}
+		}
+	}
+
+	if len(filesToRemove) > 0 {
+		c.watchesUpdated = true
+	}
+	for _, p := range filesToRemove {
+		cleaned := filepath.Dir(p)
+		// Check to see if there are paths we no longer need to watch.
+		for _, exclude := range excludedPaths {
+			excludeDir, _ := doublestar.PathMatch(exclude, cleaned)
+			if excludeDir {
+				_ = c.watcher.Remove(cleaned)
+			}
+		}
+		delete(c.watchedFiles, p)
+	}
+}
+
+// checkOnStateChanged will see if onStateChanged needs to be called.
+func (c *Component) checkOnStateChanged() {
+	if !c.watchesUpdated {
+		return
+	}
+	c.watchesUpdated = false
+	output := make([]discovery.Target, len(c.watchedFiles))
+	i := 0
+	for k := range c.watchedFiles {
+		output[i] = discovery.Target{"__path__": k}
+		i++
+	}
```

Comment:
I think this isn't quite right. It looks like that here we'd strip the targets of all their other labels, which isn't what we'd like (expand from N->M targets based on their `__path__` and leaving the other data intact).

Reply:
Fair point, if a file matches multiple paths, do we export all labels, merge, more? Goes back to earlier and lower discussion on unique by path or label.

Reply:
This is resolved now

Reply:
One thing I noticed is that we now export directories as targets, too. I'm not sure if that was the case with the fsnotify approach before.

This will make the downstream components just reject those targets, but do you think we should skip them over here as well?

Reply:
We should skip thats a bug I introduced